.github/workflows/weekly.yml

```diff
@@ -10,21 +14,53 @@ env:
   CARGO_HTTP_CHECK_REVOKE: false
 
 jobs:
+  generate-matrix:
+    runs-on: ubuntu-latest
+    outputs:
+      matrix: ${{ steps.generate-matrix.outputs.matrix }}
+    steps:
+      - uses: actions/checkout@v3
+      - uses: ./.github/actions/setup-rust
+
+      - name: Generate matrix
+        id: generate-matrix
+        run: |
+          cargo xtask ci-job target-matrix --author "[gha]" --message "check"
+        env:
+          TARGETS: ${{ inputs.targets }}
   weekly:
-    name: Check All Targets - No Cache
+    name: Check target - No Cache (${{ matrix.pretty }},${{ matrix.os }})
     timeout-minutes: 1440
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
+    needs: generate-matrix
+    strategy:
+      fail-fast: false
+      matrix:
+        include: ${{ fromJson(needs.generate-matrix.outputs.matrix) }}
     steps:
       - uses: actions/checkout@v3
       - uses: ./.github/actions/setup-rust
       - name: Set up Docker Buildx
```

Comment:
We currently only use Linux but this should probably use `if: runner.os == 'Linux'` since we aren't hard-coding `runs-on: ubuntu-latest` anymore.